File: Systems/MagicSystems/SandMasterySystem.swift
import Foundation
import SpriteKit

/// Système de Maîtrise du Sable (Sand Mastery) — Taldain
/// Les Maîtres du Sable contrôlent le sable blanc de Dayside grâce à l'Investiture solaire
final class SandMasterySystem {

    // MARK: - Sand Forms (formes de sable)

    enum SandForm: String, CaseIterable {
        case lash       // Fouet de sable — attaque directionnelle
        case shield     // Mur de sable — bouclier protecteur
        case swarm      // Essaim de sable — dégâts AoE
        case platform   // Plateforme de sable — élévation/mobilité
        case spike      // Pics de sable — piège au sol
        case storm      // Tempête de sable — ultime AoE massif
    }

    // MARK: - Utiliser une forme de sable

    struct SandResult {
        let success: Bool
        let description: String
        let damage: Int
        let healing: Int
        let statusEffect: StatusEffectType?
        let duration: TimeInterval
        let waterCost: Int  // Le sable consomme l'eau du corps
    }

    func useSandForm(
        _ form: SandForm,
        champion: inout Champion,
        targetPosition: GridPosition?
    ) -> SandResult {
        // La Maîtrise du Sable consomme l'hydratation (Investiture)
        let baseCost = 12

        guard champion.currentInvestiture >= baseCost else {
            return SandResult(success: false, description: "Déshydraté — pas assez d'eau !",
                              damage: 0, healing: 0, statusEffect: nil, duration: 0, waterCost: 0)
        }

        champion.currentInvestiture -= baseCost

        switch form {
        case .lash:
            return SandResult(
                success: true,
                description: "Fouet de Sable !",
                damage: 18 + champion.baseStats.spirit,
                healing: 0,
                statusEffect: nil,
                duration: 0,
                waterCost: baseCost
            )

        case .shield:
            return SandResult(
                success: true,
                description: "Mur de Sable Protecteur !",
                damage: 0, healing: 0,
                statusEffect: .shielded,
                duration: 6.0,
                waterCost: baseCost
            )

        case .swarm:
            return SandResult(
                success: true,
                description: "Essaim de Sable — les grains déchirent tout !",
                damage: 12 + champion.baseStats.spirit / 2,
                healing: 0,
                statusEffect: .burning,
                duration: 4.0,
                waterCost: baseCost
            )

        case .platform:
            return SandResult(
                success: true,
                description: "Plateforme de Sable — élévation !",
                damage: 0, healing: 0,
                statusEffect: .flying,
                duration: 5.0,
                waterCost: baseCost
            )

        case .spike:
            return SandResult(
                success: true,
                description: "Pics de Sable jaillissent du sol !",
                damage: 22 + champion.baseStats.spirit,
                healing: 0,
                statusEffect: .stunned,
                duration: 2.0,
                waterCost: baseCost
            )

        case .storm:
            champion.currentInvestiture -= baseCost  // Double coût pour l'ultime
            return SandResult(
                success: true,
                description: "TEMPÊTE DE SABLE — le désert se déchaîne !",
                damage: 40 + champion.baseStats.spirit * 2,
                healing: 0,
                statusEffect: .slowed,
                duration: 6.0,
                waterCost: baseCost * 2
            )
        }
    }

    // MARK: - Recharge (boire de l'eau)

    func drinkWater(champion: inout Champion, amount: Int) {
        champion.currentInvestiture = min(champion.maxInvestiture, champion.currentInvestiture + amount)
    }

    // MARK: - Effets visuels

    func sandLashEffect(from origin: CGPoint, direction: CGFloat) -> SKAction {
        SKAction.sequence([
            SKAction.moveBy(x: cos(direction) * 80, y: sin(direction) * 80, duration: 0.15),
            SKAction.removeFromParent()
        ])
    }

    func sandShieldEffect() -> SKAction {
        SKAction.repeatForever(SKAction.rotate(byAngle: .pi * 2, duration: 2.0))
    }

    func sandStormEffect() -> [SKAction] {
        (0..<8).map { i in
            let angle = CGFloat(i) * .pi / 4
            return SKAction.group([
                SKAction.moveBy(x: cos(angle) * 100, y: sin(angle) * 100, duration: 0.4),
                SKAction.fadeOut(withDuration: 0.4)
            ])
        }
    }
}

File: cosmere-web/src/game/magic/SandMasterySystem.ts
import { MagicSystem, MagicResult, emptyResult } from './MagicResult';

const SAND_COST = 12;
const STORM_COST_MULTIPLIER = 2;

interface SandForm {
  id: string;
  name: string;
  description: string;
}

const SAND_FORMS: SandForm[] = [
  { id: 'lash', name: 'Fouet de Sable', description: 'Attaque directionnelle en fouet' },
  { id: 'shield', name: 'Bouclier de Sable', description: 'Mur protecteur de sable' },
  { id: 'swarm', name: 'Essaim de Sable', description: 'Nuée de sable en zone' },
  { id: 'platform', name: 'Plateforme de Sable', description: 'Élévation — permet le vol' },
  { id: 'spike', name: 'Pics de Sable', description: 'Pièges de sable au sol' },
  { id: 'storm', name: 'Tempête de Sable', description: 'Ultime — tempête dévastatrice' },
];

export class SandMasterySystem implements MagicSystem {
  readonly name = 'Maîtrise du Sable';
  readonly worldID = 'taldain';

  private hydration = 100;

  getHydration(): number {
    return this.hydration;
  }

  setHydration(value: number): void {
    this.hydration = Math.max(0, Math.min(100, value));
  }

  drinkWater(amount: number): void {
    this.hydration = Math.min(100, this.hydration + amount);
  }

  canUseAbility(abilityID: string, currentInvestiture: number): boolean {
    const form = abilityID.replace('sand_', '');
    const sandForm = SAND_FORMS.find(f => f.id === form);
    if (!sandForm) return false;
    const cost = form === 'storm' ? SAND_COST * STORM_COST_MULTIPLIER : SAND_COST;
    return currentInvestiture >= cost && this.hydration > 0;
  }

  useAbility(abilityID: string, spirit: number): MagicResult {
    const form = abilityID.replace('sand_', '');
    const sandForm = SAND_FORMS.find(f => f.id === form);
    const result = emptyResult();

    if (!sandForm) {
      result.description = 'Forme de sable inconnue';
      return result;
    }

    const cost = form === 'storm' ? SAND_COST * STORM_COST_MULTIPLIER : SAND_COST;

    if (this.hydration <= 0) {
      result.description = 'Déshydraté — buvez de l\'eau !';
      return result;
    }

    // Sand mastery consumes hydration
    this.hydration = Math.max(0, this.hydration - (form === 'storm' ? 25 : 10));

    result.success = true;
    result.resourceCost = cost;
    result.resourceType = 'investiture';

    switch (form) {
      case 'lash':
        result.damage = 18 + spirit;
        result.description = 'Fouet de sable !';
        break;
      case 'shield':
        result.statusEffect = 'shielded';
        result.statusDuration = 6;
        result.description = 'Bouclier de sable déployé !';
        break;
      case 'swarm':
        result.damage = 12 + Math.floor(spirit / 2);
        result.statusEffect = 'burning';
        result.statusDuration = 4;
        result.areaOfEffect = true;
        result.areaRadius = 100;
        result.description = 'Essaim de sable !';
        break;
      case 'platform':
        result.statusEffect = 'flying';
        result.statusDuration = 5;
        result.description = 'Plateforme de sable — élévation !';
        break;
      case 'spike':
        result.damage = 22 + spirit;
        result.statusEffect = 'stunned';
        result.statusDuration = 2;
        result.description = 'Pics de sable jaillis du sol !';
        break;
      case 'storm':
        result.damage = 40 + spirit * 2;
        result.statusEffect = 'slowed';
        result.statusDuration = 6;
        result.areaOfEffect = true;
        result.areaRadius = 200;
        result.description = 'Tempête de sable déchaînée !';
        break;
    }

    return result;
  }

  getAvailableAbilities(): string[] {
    return SAND_FORMS.map(f => `sand_${f.id}`);
  }
}
